drv/sidecar-front-io/src/transceivers.rs

```diff
@@ -0,0 +1,201 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+use crate::{Addr, Reg};
+use drv_fpga_api::{FpgaError, FpgaUserDesign, WriteOp};
+use zerocopy::{AsBytes, FromBytes, Unaligned};
+
+pub struct Transceivers {
+    fpgas: [FpgaUserDesign; 2],
+}
+
+impl Transceivers {
+    pub fn new(fpga_task: userlib::TaskId) -> Self {
+        Self {
+            // There are 16 QSFP-DD transceivers connected to each FPGA
+            fpgas: [
+                FpgaUserDesign::new(fpga_task, 0),
+                FpgaUserDesign::new(fpga_task, 1),
+            ],
+        }
+    }
+
+    pub fn get_modules_status(&self) -> Result<[u32; 7], FpgaError> {
+        let mut r: [u32; 7] = [0; 7];
+        let f0: [u8; 14] = self.fpgas[0].read(Addr::QSFP_CTRL_EN_H)?;
+        let f1: [u8; 14] = self.fpgas[1].read(Addr::QSFP_CTRL_EN_H)?;
+
+        for i in 0..7 {
+            r[i] = ((f1[i * 2] as u32) << 24)
+                | ((f1[i * 2 + 1] as u32) << 16)
+                | ((f0[i * 2] as u32) << 8)
+                | (f0[i * 2 + 1] as u32);
+        }
```

Comment:
This is fine, but a bit bit-twiddly and hard to follow. If I understand the register map a slightly more functional/Rust approach, which expresses the idea of stitching together two half of an array, might be something along the lines of
```
let q0_q15: [u16; 7] = self.fpgas[0].read(Addr::QSFP_CTRL_EN_H)?;
let q16_q31: [u16; 7] = self.fpgas[1].read(Addr::QSFP_CTRL_EN_H)?;

let status = q0_q15.iter()
    .zip(q16_q31.iter())
    .map(|(q0_q15, q16_q31)| {
        ((q16_q31 as u32) << 16) | (q0_q15 as u32)
    });
    
Ok(status)
```
Depending on byte ordering you may have to do something like
```
((u16::from_be(q16_q31) as u32) << 16) | (u16::from_be(q0_q15) as u32)
```

Reply:
You could remove one layer of combination using `zerocopy::{U16, BE}` to read into U16s directly, then combine them with `zip`:
```rust
        let f0: [U16<BE>; 7] = self.fpgas[0].read(Addr::QSFP_CTRL_EN_H)?;
        let f1: [U16<BE>; 7] = self.fpgas[1].read(Addr::QSFP_CTRL_EN_H)?;

        let mut out: [u32; 7] = [0; 7];
        for (out, (lo, hi)) in out.iter_mut().zip(f0.iter().zip(f1.iter())) {
            *out = (lo.get() as u32) | ((hi.get() as u32) << 16);
        }

        Ok(out)
```
(you'll also need to add `U16` and `BE` to the `zerocopy` imports at the top)

More broadly, is there a reason this function wants to return raw bytes instead of `ModulesStatus`?